### ISSUE TITLE
Unable to connect via doh on non-443 HTTPS port

### ISSUE THREAD

**Issue by seanob86** [DNSCrypt/dnscrypt-proxy#888]
I have configured my own Adguard Home VPS box and successfully enabled the encryption settings in Adguard Home to allow dnscrypt-proxy (2.0.25) clients to connect to it via doh. In my toml config file I simply uncommented the static server section and provided the doh stamp of my server.

This all works well when the HTTPS port in AdGuard Home is `443`. However, if I change the HTTPS port from 443 to (as an example) 8081 and update the stamp in my toml config file, dnscrypt-proxy is unable to connect. Here is the error - note I have removed my hostname from the log

```
[2019-07-18 19:01:40] [NOTICE] Network connectivity detected
[2019-07-18 19:01:40] [NOTICE] Source [public-resolvers.md] loaded
[2019-07-18 19:01:40] [NOTICE] dnscrypt-proxy 2.0.25
[2019-07-18 19:01:40] [NOTICE] Loading the set of blocking rules from [/usr/local/opt/dnscrypt-proxy/sbin/blacklist.txt]
[2019-07-18 19:01:40] [NOTICE] Loading the set of forwarding rules from [/usr/local/opt/dnscrypt-proxy/sbin/forwarding-rules.txt]
[2019-07-18 19:01:40] [NOTICE] Now listening to 127.0.0.1:53 [UDP]
[2019-07-18 19:01:40] [NOTICE] Now listening to 127.0.0.1:53 [TCP]
[2019-07-18 19:01:40] [NOTICE] Now listening to [::1]:53 [UDP]
[2019-07-18 19:01:40] [NOTICE] Now listening to [::1]:53 [TCP]
[2019-07-18 19:01:45] [ERROR] Get https://[hostname-removed]:8081/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[2019-07-18 19:01:45] [NOTICE] dnscrypt-proxy is waiting for at least one server to be reachable
```

On the VPS server I have cleared all the iptables to `ACCEPT` all incoming, outgoing and forward connections.

The only way I can manage dnscrypt-proxy to successfully connect to my adguard home server (via custom HTTPS port) is by adding an additional public DNS server on my laptop network DNS settings. So by going from just `127.0.0.1`, I updated to `127.0.0.1` and `1.1.1.1`. Then trying dnscrypt-proxy to connect will establish a connection to my AdGuard Home Server. i can even see in the logs in adguard home my laptop (hosting dnscrypt-proxy) has connected via the custom HTTPS port.

Is there a way for all this to work successfully with just 127.0.0.1?

For reference my laptop (with dnscrypt-proxy installed) is macOS but I have also tested on my Raspberry Pi with the same error.

Also, screenshoot of the option in Adguard Home to modify HTTPS port
<img width="461" alt="Screen Shot 2019-07-18 at 19 40 19" src="https://user-images.githubusercontent.com/45988261/61451229-778d5000-a948-11e9-8d2c-aa326f5a6e41.png">

**Comment by jedisct1** [comment]
Can you connect to that host name and port with a web browser or `nc`?

**Comment by jedisct1** [comment]
Since the `GET` query is displayed, you can directly try the complete URL:

```sh
https://[hostname-removed]:8081/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA
```

**Comment by seanob86** [comment]
I can. As an example https://myhostname:8081. Doing just that will present me with the Adguard Home webpage. 

EDIT: if URL entered in browser. I can also try the URL get query as well soon.

**Comment by seanob86** [comment]
OK, so i'm not sure how to run the `nc` command however.. when I use the get query url (below) via my web-browser i am prompted to download a dns-query file (which I can do successfully).

`https://[hostname-removed]:8081/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA`

EDIT: if I run `curl -I https://[hostname]:8081` I get the following
```

HTTP/2 401 
access-control-allow-origin: *
www-authenticate: Basic realm="dnsfilter"
content-type: text/plain; charset=utf-8
content-length: 14
date: Thu, 18 Jul 2019 12:48:02 GMT
```

**Comment by publicarray** [comment]
The 401 status code means “unauthorised” and the www-authenticate header suggest that you have “basic authentication” enabled on that URL. Can you disable it temporarily and see if that is the problem. As far as i know dnscrypt-proxy can’t deal with authenticated endpoints. Maybe creating a proxy or VPN to your server might do what you want.

**Comment by seanob86** [comment]
> The 401 status code means “unauthorised” and the www-authenticate header suggest that you have “basic authentication” enabled on that URL. Can you disable it temporarily and see if that is the problem. As far as i know dnscrypt-proxy can’t deal with authenticated endpoints. Maybe creating a proxy or VPN to your server might do what you want.

Apologies - it appears i ran the curl command without '/dns-query' at the end of the url. The authentication is the AdGuard Home UI login. I will provide output for both port `443` and `8081` below with '/dns-query'. Note, port `443` has no issues connecting but `8081` refuses to connect.

**Port 443**
```

$ curl -I https://[hostname]/dns-query
HTTP/2 405 
access-control-allow-origin: *
content-type: text/plain; charset=utf-8
x-content-type-options: nosniff
content-length: 19
date: Tue, 23 Jul 2019 01:18:05 GMT

```
**Port 8081**
```

$ curl -I https://[hostname]:8081/dns-query
HTTP/2 405 
access-control-allow-origin: *
content-type: text/plain; charset=utf-8
x-content-type-options: nosniff
content-length: 19
date: Tue, 23 Jul 2019 01:23:32 GMT

```

**Comment by publicarray** [comment]
Ah no-worries can you please try this with your hostname instead? and copy the output? This will provide more debugging information.

query for example.com:

`time curl -H 'accept: application/dns-message' -v 'https://cloudflare-dns.com/dns-query?dns=AAABAAABAAAAAAAAA3d3dwdleGFtcGxlA2NvbQAAAQAB' | hexdump -C`

query for root:

`time curl -H 'accept: application/dns-message' -v 'https://cloudflare-dns.com/dns-query?dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA' | hexdump -C`

It looks like a service is running on both ports, at least it's responding with a http response.

If you like you can send me [DM with the output via twitter](https://twitter.com/messages/compose?recipient_id=765886034106351616). And you can use something like https://privatebin.net/ or https://send.firefox.com/ to paste the content.

**Comment by publicarray** [comment]
Just for reference here is a working stamp that has a different port number for DoH
`sdns://AgcAAAAAAAAADDQ1Ljc2LjExMy4zMSA-GhoPbFPz6XpJLVcIS1uYBwWe4FerFQWHb9g_2j24OBBkb2guc2VieS5pbzo4NDQzCi9kbnMtcXVlcnk`

You can paste it in here: https://dnscrypt.info/stamps

**Comment by seanob86** [comment]
> query for example.com:
> 
> time curl -H 'accept: application/dns-message' -v 'https://cloudflare-dns.com/dns-query?dns=AAABAAABAAAAAAAAA3d3dwdleGFtcGxlA2NvbQAAAQAB' | hexdump -C
> 
> query for root:
> 
> time curl -H 'accept: application/dns-message' -v 'https://cloudflare-dns.com/dns-query?dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA' | hexdump -C

Hey @publicarray - I will send you a DM via twitter with the command outputs with port 8081. From what I can tell it looks fine.

> Just for reference here is a working stamp that has a different port number for DoH
sdns://AgcAAAAAAAAADDQ1Ljc2LjExMy4zMSA-GhoPbFPz6XpJLVcIS1uYBwWe4FerFQWHb9g_2j24OBBkb2guc2VieS5pbzo4NDQzCi9kbnMtcXVlcnk
```

[2019-07-25 10:12:30] [ERROR] Get https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[2019-07-25 10:12:30] [NOTICE] dnscrypt-proxy is waiting for at least one server to be reachable

```
This also did not work for me with **only** 127.0.0.1 in my network DNS.
I could only connect to https://doh.seby.io:8443 if I add a secondary DNS server (see below).

This is the same behavior on my private VPS AdGuard Home server on any port other than 443. It's as though it can't resolve the hostname - which is strange cos it works with port 443.

<img width="780" alt="Screen Shot 2019-07-25 at 10 36 07" src="https://user-images.githubusercontent.com/45988261/61839305-9b86ef00-ae7c-11e9-8876-78400838fbfd.png">

**Comment by publicarray** [comment]
Your server is working fine maybe try changing the fallback resolver on the client to `fallback_resolver = '1.1.1.1:53'`

@jedisct1
The response via curl was within 0.2s and the response was valid.

**Comment by publicarray** [comment]
You can also change the ignore_system_dns setting `ignore_system_dns = false` and see if true or false works

**Comment by publicarray** [comment]
Never-mind I can reproduce the error too.

`sudo env DEBUG=1 dnscrypt-proxy -loglevel 0 -config /usr/local/etc/dnscrypt-proxy.toml`

```
[2019-07-25 19:22:34] [NOTICE] dnscrypt-proxy 2.0.25
[2019-07-25 19:22:34] [NOTICE] Loading the set of blocking rules from [/usr/local/etc/dnscrypt-proxy/blacklist.txt]
[2019-07-25 19:22:34] [NOTICE] Loading the set of cloaking rules from [/usr/local/etc/dnscrypt-proxy/cloaking-rules.txt]
[2019-07-25 19:22:34] [NOTICE] Now listening to 127.0.0.1:53 [UDP]
[2019-07-25 19:22:34] [NOTICE] Now listening to 127.0.0.1:53 [TCP]
[2019-07-25 19:22:34] [NOTICE] Now listening to [::1]:53 [UDP]
[2019-07-25 19:22:34] [NOTICE] Now listening to [::1]:53 [TCP]
[2019-07-25 19:22:34] [DEBUG] Refreshing certificates
[2019-07-25 19:22:34] [DEBUG] [doh.seby.io] IP address was not cached
[2019-07-25 19:22:34] [DEBUG] clients count: 1
[2019-07-25 19:22:34] [DEBUG] clients count: 1
[2019-07-25 19:22:35] [DEBUG] clients count: 1
[2019-07-25 19:22:37] [DEBUG] [https://doh.seby.io:8443/dns-query?body_hash=fc3f9e09b65a7ce1e47b02b2991f5934eccb5bca444ef8619f541537f636529a]: [Post https://doh.seby.io:8443/dns-query?body_hash=fc3f9e09b65a7ce1e47b02b2991f5934eccb5bca444ef8619f541537f636529a: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)]
[2019-07-25 19:22:37] [DEBUG] [doh.seby.io] IP address was not cached
[2019-07-25 19:22:37] [DEBUG] clients count: 1
[2019-07-25 19:22:39] [DEBUG] [https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA]: [Get https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)]
[2019-07-25 19:22:39] [ERROR] Get https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[2019-07-25 19:22:39] [NOTICE] dnscrypt-proxy is waiting for at least one server to be reachable
[2019-07-25 19:22:41] [DEBUG] clients count: 1
[2019-07-25 19:22:47] [DEBUG] clients count: 1
[2019-07-25 19:22:47] [DEBUG] clients count: 2
[2019-07-25 19:22:49] [DEBUG] clients count: 1
[2019-07-25 19:22:49] [DEBUG] Refreshing certificates
[2019-07-25 19:22:49] [DEBUG] [doh.seby.io] IP address was not cached
[2019-07-25 19:22:52] [DEBUG] [https://doh.seby.io:8443/dns-query?body_hash=fc3f9e09b65a7ce1e47b02b2991f5934eccb5bca444ef8619f541537f636529a]: [Post https://doh.seby.io:8443/dns-query?body_hash=fc3f9e09b65a7ce1e47b02b2991f5934eccb5bca444ef8619f541537f636529a: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)]
[2019-07-25 19:22:52] [DEBUG] [doh.seby.io] IP address was not cached
[2019-07-25 19:22:54] [DEBUG] [https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA]: [Get https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)]
[2019-07-25 19:23:04] [DEBUG] Refreshing certificates
[2019-07-25 19:23:04] [DEBUG] [doh.seby.io] IP address was not cached
[2019-07-25 19:23:04] [DEBUG] clients count: 1
[2019-07-25 19:23:05] [DEBUG] clients count: 1
[2019-07-25 19:23:07] [DEBUG] [https://doh.seby.io:8443/dns-query?body_hash=fc3f9e09b65a7ce1e47b02b2991f5934eccb5bca444ef8619f541537f636529a]: [Post https://doh.seby.io:8443/dns-query?body_hash=fc3f9e09b65a7ce1e47b02b2991f5934eccb5bca444ef8619f541537f636529a: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)]
[2019-07-25 19:23:07] [DEBUG] [doh.seby.io] IP address was not cached
[2019-07-25 19:23:07] [DEBUG] clients count: 1
[2019-07-25 19:23:09] [DEBUG] [https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA]: [Get https://doh.seby.io:8443/dns-query?ct=&dns=yv4BAAABAAAAAAABAAACAAEAACkQAAAAgAAAAA: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)]
[2019-07-25 19:23:11] [DEBUG] clients count: 1
```

cc/ @jedisct1

Edit: changing the `fallback_resolver` or `ignore_system_dns` makes no difference

**Comment by publicarray** [comment]
So the problem is that dnscrypt-proxy is not able to bootstrap the DoH server correctly or is unable to correctly use the stamp with a port number and IP. `sdns://AgcAAAAAAAAADDQ1Ljc2LjExMy4zMSA-GhoPbFPz6XpJLVcIS1uYBwWe4FerFQWHb9g_2j24OBBkb2guc2VieS5pbzo4NDQzCi9kbnMtcXVlcnk` It's obviously asking the OS where doh.seby.io is.

**Comment by seanob86** [comment]
I just finished my testing and have come to the same conclusion regarding the fallback resolver! 

I also have more info.

So in my toml file I have uncommented `server_names` so dnscrypt-proxy only connects to my server. Mine looks like this: `server_names = ['seanob-agh']`

However, if I **comment** this line so that dnscrypt-proxy filters through all the public servers (including my private server), it manages to connect fine but the result is that dnscrypt-proxy chooses a server with the lowest ping time which ultimately is not my server. 
Here is the successful connection below where `server_names` is commented
```
Jul 25 19:24:28 raspberrypi dnscrypt-proxy[1027]: [2019-07-25 19:24:28] [NOTICE] [seanob-agh] OK (DoH) - rtt: 42ms

```
Then un-commenting` server_names` will fail to connect to my desired server (on linux). On Mac it manages to connect with server_names uncommented however when browsing fails to load any sites in my browser I try restarting the service and then i'm back to square one.

Hope this troubleshooting helps!
